File: ports/ts/src/lang/lang.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import { resolveAnswerLanguage, type LanguageResult } from "./lang.js";

// Proven against the shared fixture — every case must match the Python reference
// citenexus.lang.fallback.resolve_answer_language exactly. Follows the §4
// exemplar: load conformance/cases/language.json, assert equality over ALL
// cases, no leniency.
interface LangCase {
  name: string;
  detection: LanguageResult | null;
  answer_language: string | null;
  conversation_language: string | null;
  languages_in_evidence: string[];
  default_answer_language: string;
  expected: string;
}

describe("resolve_answer_language conformance", () => {
  const cases = loadCase<LangCase[]>("language.json");

  it("has cases", () => {
    expect(cases.length).toBeGreaterThan(0);
  });

  for (const c of cases) {
    it(c.name, () => {
      expect(
        resolveAnswerLanguage({
          detection: c.detection,
          answer_language: c.answer_language,
          conversation_language: c.conversation_language,
          languages_in_evidence: c.languages_in_evidence,
          default_answer_language: c.default_answer_language,
        }),
      ).toEqual(c.expected);
    });
  }
});

File: golang/lang/lang_test.go
package lang

import (
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// Proven against the shared fixture: every §11a fallback case must match the
// Python reference resolve_answer_language exactly, no leniency.
// expectedLanguageCases pins the size of conformance/cases/language.json.
const expectedLanguageCases = 6

func TestLanguageVectorCount(t *testing.T) {
	var cases []struct {
		Name string `json:"name"`
	}
	conform.Case(t, "language.json", &cases)
	if len(cases) != expectedLanguageCases {
		t.Fatalf("language.json: got %d cases, want %d", len(cases), expectedLanguageCases)
	}
}

func TestResolveAnswerLanguageConformance(t *testing.T) {
	var cases []struct {
		Name                 string     `json:"name"`
		Detection            *Detection `json:"detection"`
		AnswerLanguage       string     `json:"answer_language"`
		ConversationLanguage string     `json:"conversation_language"`
		LanguagesInEvidence  []string   `json:"languages_in_evidence"`
		DefaultAnswerLang    string     `json:"default_answer_language"`
		Expected             string     `json:"expected"`
	}
	conform.Case(t, "language.json", &cases)

	if len(cases) != expectedLanguageCases {
		t.Fatalf("language.json: got %d cases, want %d", len(cases), expectedLanguageCases)
	}
	for _, c := range cases {
		got := ResolveAnswerLanguage(
			c.Detection,
			c.AnswerLanguage,
			c.ConversationLanguage,
			c.LanguagesInEvidence,
			c.DefaultAnswerLang,
		)
		if got != c.Expected {
			t.Errorf("%s: ResolveAnswerLanguage = %q, want %q", c.Name, got, c.Expected)
		}
	}
}
